### PR TITLE
Add downloadJdbcDrivers function and remove dependency on DatabaseConnectorJars package.

### PR DESCRIPTION
**Pull request by ablack3**
DatabaseConnectorJars is now too large for CRAN. An alternative approach is to have the user download the jar files and put them in a location where they can be accessed by the DatabaseConnector package. To this end a new function is proposed in this pull request which downloads the jar files for the user. Furthermore we propose using the environment variable DATABASECONNECTOR_JAR_FOLDER to store the path to the jar files.

I'm not sure about all the implications of these changes so I'm creating a draft PR to get feedback and suggestions.

### CODE REVIEW

File: R/Connect.R [hunk 1]
```diff
@@ -82,6 +82,57 @@ createConnectionDetails <- function(dbms,
 }
 
 jdbcDrivers <- new.env()
+jdbcDrivers$defaultPath <- Sys.getenv("DATABASECONNECTOR_JAR_FOLDER", 
+                                      unset = path.expand("~/DatabaseConnectorDrivers"))
```
Comment (schuemie) [inline R/Connect.R:86]: Why store an environmental variable in a global variable? This way updates to the environmental variable will have no impact. Can't we just always refer to the environmental variable throughout the code?

File: R/Connect.R [hunk 2]
```diff
@@ -82,6 +82,57 @@ createConnectionDetails <- function(dbms,
...
+#' 
+#' @return Invisibly returns the destination if the download was successful.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' downloadJdbcDrivers()
+#' }
+downloadJdbcDrivers <- function(destination = NULL, method = "auto", ...){
+  if (is.null(destination) || missing(destination)) {
```
Comment (schuemie) [inline R/Connect.R:106]: Should we allow the user to specify which JDBC driver to download? Likely they'll only need one.

File: R/Connect.R [hunk 3]
```diff
@@ -82,6 +82,57 @@ createConnectionDetails <- function(dbms,
...
+  }
+  
+  if (!dir.exists(destination)) dir.create(destination)
+  
+  baseUrl <- "https://github.com/OHDSI/DatabaseConnectorJars/blob/master/inst/java"
+  jdbcDriverNames <- c("RedshiftJDBC4-no-awssdk-1.2.20.1043.jar",
+                       "ojdbc8.jar",
+                       "postgresql-42.2.5.jre6.jar",
+                       "sqljdbc4.jar")
+
```
Comment (schuemie) [inline R/Connect.R:117]: RedShift now requires a [ton of jars](https://github.com/OHDSI/DatabaseConnectorJars/tree/develop/inst/java) to be installed (which is why we're being refused by CRAN). We need to install all of those (which also argues for allowing users to specify which driver to install).

**Review by schuemie — COMMENTED** [review]
Thanks! I'm not yet sure how we'd serve the jars for people to download. Perhaps a more elegant option than pointing to a folder in a (soon to be obsolete) package is to host the jars in GitHub Pages of DatabaseConnectorJars?

**Comment by schuemie** [comment]
I've placed the JDBC drivers on the [GitHub pages of DatabaseConnectorJars](https://ohdsi.github.io/DatabaseConnectorJars/index.html). We can have DatabaseConnector download from here. I've clearly marked the versions (e.g. 'oracleV19.8.zip') because possibly a specific version of DatabaseConnector will only work with a specific version of the JDBC driver, so DatabaseConnector should specify which version it downloads. Note that some driver zip files (i.e. RedShift's) contain multiple jars.

File: R/Connect.R (hunk 1 above)
Comment (ablack3) [reply R/Connect.R:86]: Yes we can. I want to provide a default path in the event that the environment variable is not set. I'll make the change

File: R/Connect.R (hunk 2 above)
Comment (ablack3) [reply R/Connect.R:106]: Yes. Agreed.

File: R/Connect.R (hunk 3 above)
Comment (ablack3) [reply R/Connect.R:117]: image
Ok. I'm assuming all of the new jar files are for Redshift. Maybe we should have a readme file in this folder explaining what the files are and where they came from.

File: R/Connect.R (hunk 1 above)
Comment (schuemie) [reply R/Connect.R:86]: I'm pretty sure CRAN will not allow us to default to a path in the user's folder, so I recommend the following: There are I think 3 places where the path to the jars must be specified:
1. downloadJdbcDrivers()
2. connect()
3. createConnectionDetails()

I recommend each of these just get a `pathToDriver` argument (the latter two already have that), that defaults to the environmental variable. So the user can choose to set the folder manually when calling the function, or setting it in the environmental variable. If no folder is specified, a meaningful error should be thrown.

**Comment by ablack3** [comment]
> I've clearly marked the versions (e.g. 'oracleV19.8.zip') because possibly a specific version of DatabaseConnector will only work with a specific version of the JDBC driver, so DatabaseConnector should specify which version it downloads.

How should the driver version be specified in DatabaseConnector?
Will only the latest version be available on Github Pages?
I'm unsure how driver version managment should work and can see things possibly not working well if I just hardcode the version (e.g. 'oracleV19.8.zip') into the downloadDrivers() function.

File: R/Connect.R [hunk 4]
```diff
@@ -110,7 +183,7 @@ getJbcDriverSingleton <- function(driverClass = "", classPath = "") {
 
 findPathToJar <- function(name, pathToDriver) {
   if (missing(pathToDriver) || is.null(pathToDriver)) {
-    pathToDriver <- system.file("java", package = "DatabaseConnectorJars")
+    pathToDriver <- jdbcDrivers$defaultPath
```
Comment (schuemie) [inline R/Connect.R:186]: Does `jdbcDrivers$defaultPath` have a value? I think this code is now obsolete?

File: R/Connect.R (hunk 3 above)
Comment (ablack3) [reply R/Connect.R:137]: The user can specify which dbms they need drivers for.

File: R/Connect.R (hunk 1 above)
Comment (ablack3) [reply R/Connect.R]: I think I've made the necessary changes according to your suggestions.

File: R/Connect.R (hunk 4 above)
Comment (ablack3) [reply R/Connect.R:186]: Yes. You're correct. I don't have a good way to run the automated tests for this package yet. I need to get that set up which will help me catch errors before making commits!

Comment (ablack3) [reply R/Connect.R]: I think this is fixed but I have not tested the code in this PR. Hopefully I can get that set up soon. I might need to modify some of the tests by breaking them by database. (sql server, postgres, oracle, etc,  would be in separate tests so I can just run the tests on the databases I have access to.)

**Comment by schuemie** [comment]
Excellent! I think this should do it. In response to your questions:

I'm not sure what the right approach to picking the version of the JDBC driver to download is. I'm leaning towards us hardcoding that in DatabaseConnector, so that a specific version of DatabaseConnector corresponds to a specific version of a JDBC driver. That way we can test the combination and verify it works. What do you think? (A related point: one major difference between DatabaseConnector and RJDBC is that the latter tries to be driver-agnostic, while the former tries to actually work in practice with specific drivers and their idiosyncrasies. I've learned over the years that the two goals are mutually exclusive) 

I'm also not sure whether all drivers should be in one folder. That is certainly the easiest thing to program, and people will likely have only one version of one JDBC driver on their system. In theory, if people use renv they might be running different versions of DatabaseConnector on one system, and as mentioned above each would have its own specific version of the JDBC driver, so that would argue for folders per version. Your thoughts?

**Comment by ablack3** [comment]
Great! I'll remove the "draft" status and try to resolve the conflicts.

I agree we should just hardcode the driver versions into DatabaseConnector but maybe we should put the names of the drivers in some central place so that they are only coded once in package. Perhaps we should put the driver names in the jdbcDrivers environment.

Thanks for the info about the difference between DatabaseConnector and RJDBC. I see the benefit of only using a small set of drivers but still think it is worth exploring decoupling DatabaseConnector's query functions from specific JDBC drivers. If we stick to the DBI interface then it seems like any DBI compliant connection should be able to work with the OHDSI tools. I just think it is quite limiting to tie all of OHDSI's analytic tools to specific drivers. I will think about how this might be done and create a proposal for discussion. (An optional BYOD user story - bring your own driver.) 

For storing the drivers, I'm not sure about the best approach either. I tend to think keeping the redshift jars in a separate subfolder might alleviate confusion since it will be clear that they are all required for redshift. I'll take a stab at implementing the subfolder method.

**Pull request merged**